src/xen/xennet.c

```diff
@@ -0,0 +1,778 @@
+    if (notify)
+        xen_notify_evtchn(xd->evtchn);
+}
+
+static CLOSURE_2_0(xennet_rx_buf_finish, void, xennet_dev, struct pbuf *);
+static void xennet_rx_buf_finish(xennet_dev xd, struct pbuf *p)
+{
+    if (xd->netif->input(p, xd->netif) != ERR_OK) {
+        msg_err("rx drop by stack\n");
+        xennet_return_rxbuf(p);
+    }
+}
+
+static void xennet_service_rx_ring(xennet_dev xd)
+{
+    int more;
+
+    do {
+        RING_IDX cons = xd->rx_ring.rsp_cons;
+        RING_IDX prod = xd->rx_ring.sring->rsp_prod;
+        assert(prod - cons < XENNET_RX_RING_SIZE);
+        read_barrier();
+        xennet_debug("%s: cons %d, prod %d", __func__, cons, prod);
+
+        while (cons < prod) {
+            netif_rx_response_t *rx = RING_GET_RESPONSE(&xd->rx_ring, cons);
+            xennet_rx_buf rxb = vector_get(xd->rxbufs, rx->id);
+            assert(rxb);
+            assert(rxb->gntref != GRANT_INVALID);
+            xen_revoke_page_access(rxb->gntref);
+            rxb->gntref = GRANT_INVALID;
+
+            assert(rx->status >= 0);
+            assert(rx->offset + rx->status <= PAGESIZE);
+
+            xennet_debug("   RX flags %x, status %d, offset %d, buf:\n%X",
+                         rx->flags, rx->status, rx->offset,
+                         alloca_wrap_buffer(rxb->p.pbuf.payload, rx->status + rx->offset));
+            rxb->p.pbuf.len = rx->status;
+            rxb->p.pbuf.tot_len = rx->status;
+            rxb->p.pbuf.payload += rx->offset;
+
+            enqueue(bhqueue, closure(xd->h, xennet_rx_buf_finish, xd, &rxb->p.pbuf));
+            cons++;
+        }
+        write_barrier();
+        xd->rx_ring.rsp_cons = cons;
+        RING_FINAL_CHECK_FOR_RESPONSES(&xd->rx_ring, more);
+    } while (more);
+}
+
+static CLOSURE_1_0(xennet_event_handler, void, xennet_dev);
+static void xennet_event_handler(xennet_dev xd)
+{
+    xennet_service_tx_ring(xd);
+    xennet_populate_tx_ring(xd);
+    xennet_service_rx_ring(xd);
+    xennet_populate_rx_ring(xd);
+    int rv = xen_unmask_evtchn(xd->evtchn);
+    if (rv != 0)
+        halt("%s: failed to unmask evtchn %d, rv %d\n", xd->evtchn, rv);
+}
+
+static status xennet_enable(xennet_dev xd)
+{
+    status s = STATUS_OK;
+    xennet_debug("%s: dev id %d", __func__, xd->if_id);
+
+    u64 val;
+    s = xenstore_read_u64(0, xd->backend, "feature-rx-copy", &val);
+    if (!is_ok(s)) {
+        s = timm("result", "failed to verify presence of rx-copy feature: %v", s);
+        return s;
+    }
+
+    if (!val) {
+        s = timm("result", "rx-copy not supported by backend");
+        return s;
+    }
+
+    xd->rx_ring_gntref = GRANT_INVALID;
+    xd->tx_ring_gntref = GRANT_INVALID;
+
+    /* allocate shared rings */
+    netif_rx_sring_t *rx_ring = allocate_zero(xd->contiguous, PAGESIZE);
+    netif_tx_sring_t *tx_ring = allocate_zero(xd->contiguous, PAGESIZE);
+    assert(rx_ring != INVALID_ADDRESS);
+    assert(tx_ring != INVALID_ADDRESS);
+    
+    SHARED_RING_INIT(rx_ring);
```

Comment:
Do these rings not also need to be manipulated via volatile pointers? I know this is public code, but just curious why the other hypervisor interfaces (in `xen_platform_info`) are declared volatile and these aren't

Reply:
This is a good question. In the case of these rings, I believe it's not necessary to use volatile because the macros which push the ring changes (e.g. RING_PUSH_REQUESTS_AND_CHECK_NOTIFY, RING_FINAL_CHECK_FOR_RESPONSES) use the various memory barrier macros, which in turn specify the "memory" clobber in their respective asm inlines. Memory barriers are also used prior to the ring processing loops as well. Neither Linux nor NetBSD uses volatiles for the rings. I suspect it would be advantageous for performance to avoid hitting memory with each read/write within these loops as well; the pattern here is to only update the ring pointers after each linear run of processing ring elements. The other shared page accesses mentioned above are akin to device register accesses, where it's safer to avoid stashing values in registers and instead always hit memory/cache with each access.